javalin/src/main/java/io/javalin/core/validation/BaseValidator.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Javalin - https://javalin.io
+ * Copyright 2017 David Åse
+ * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
+ */
+
+package io.javalin.core.validation
+
+import io.javalin.http.BadRequestResponse
+
+typealias Check<T> = (T) -> Boolean
+typealias ErrorProvider = () -> String
+
+data class Rule<T>(val fieldName: String, val check: Check<T?>, val errorProvider: ErrorProvider)
+
+open class BaseValidator<T>(val value: T?) {
+
+    private val rules = mutableSetOf<Rule<T>>()
+
+    fun addRule(fieldName: String, check: Check<T?>, errorProvider: ErrorProvider): BaseValidator<T> {
+        rules.add(Rule(fieldName, check, errorProvider))
+        return this
+    }
+
+    open fun get(): T? = when {
+        rules.all { it.check(value) } -> value // all rules valid
+        else -> throw BadRequestResponse(rules.firstErrorMsg(value) ?: ValidationError.CUSTOM_CHECK_FAILED.name)
```

Comment:
Is there a reason not to just call errors in this function and then check if the map is empty? It'd be nice to have the error thrown from here actually contain all the errors, as it's generally frustrating to a user to have to fix errors on a form one item at a time. 

Additionally, while this kind of fails fast, it also goes through everything up to and including the first error twice, and those checks could be expensive for checks that are making sure an email address isn't already taken.

If you want to fail fast you could do `rules.find { !it.check(value) }?.run { throw BadRequestResponse(it...) } ?: value`

Reply:
> Is there a reason not to just call errors in this function and then check if the map is empty?

Not really, this is just how it was. I like your suggestion.